sbat.c

```diff
@@ -117,4 +118,191 @@ parse_sbat(char *sbat_base, size_t sbat_size, struct sbat *sbat)
+{
+	UINT16 sbat_gen, sbat_var_gen;
+
+	if (strcmp(entry->component_name, sbat_var_entry->component_name) ==
+	    0) {
+		dprint(L"component %a has a matching SBAT variable entry, verifying\n",
+			entry->component_name);
+
+		/*
+		 * atoi returns zero for failed conversion, so essentially
+		 * badly parsed component_generation will be treated as zero
+		 */
+		sbat_gen = atoi(entry->component_generation);
+		sbat_var_gen = atoi(sbat_var_entry->component_generation);
+
+		if (sbat_gen < sbat_var_gen) {
+			dprint(L"component %a, generation %d, was revoked by SBAT variable",
+			       entry->component_name, sbat_gen);
+			LogError(L"image did not pass SBAT verification\n");
+			return EFI_SECURITY_VIOLATION;
+		}
+	}
+	return EFI_SUCCESS;
+}
+
+static void
+clean_up_vars(list_t *entries)
+{
+	list_t *pos = NULL, *tmp = NULL;
+	struct sbat_var *entry;
+
+	for_each_sbat_var_safe(pos, tmp, entries)
+	{
+		entry = list_entry(pos, struct sbat_var, list);
+		list_del(&entry->list);
+
+		if (entry->component_generation)
+			FreePool((CHAR8 *)entry->component_name);
+		if (entry->component_name)
+			FreePool((CHAR8 *)entry->component_generation);
+		FreePool(entry);
+	}
+}
+
+EFI_STATUS
+verify_sbat(struct sbat *sbat, list_t *sbat_entries)
+{
+	unsigned int i;
+	list_t *pos = NULL;
+	EFI_STATUS efi_status = EFI_SUCCESS;
+	struct sbat_entry *entry;
+	struct sbat_var *sbat_var_entry;
+
+	if (!sbat_entries || sbat_empty(sbat_entries)) {
+		dprint(L"SBAT variable not present or malformed\n");
+		return EFI_INVALID_PARAMETER;
+	}
+
+	for (i = 0; i < sbat->size; i++) {
+		entry = sbat->entries[i];
+		for_each_sbat_var(pos, sbat_entries)
+		{
+			sbat_var_entry = list_entry(pos, struct sbat_var, list);
+			efi_status = verify_single_entry(entry, sbat_var_entry);
+			if (EFI_ERROR(efi_status))
+				return efi_status;
+		}
+	}
+
+	dprint(L"all entries from SBAT section verified\n");
+	clean_up_vars(sbat_entries);
+	return efi_status;
+}
+
+static BOOLEAN
+is_utf8_bom(CHAR8 *buf, size_t bufsize)
+{
+	unsigned char bom[] = { 0xEF, 0xBB, 0xBF };
+
+	return CompareMem(buf, bom, MIN(sizeof(bom), bufsize)) == 0;
+}
+
+static struct sbat_var *
+new_entry(const CHAR8 *comp_name, const CHAR8 *comp_gen)
+{
+	struct sbat_var *new_entry = AllocatePool(sizeof(*new_entry));
+
+	if (!new_entry)
+		return NULL;
+	INIT_LIST_HEAD(&new_entry->list);
+	new_entry->component_name = comp_name;
+	new_entry->component_generation = comp_gen;
+
+	return new_entry;
+}
+
+EFI_STATUS
+add_entry(list_t *list, const CHAR8 *comp_name, const CHAR8 *comp_gen)
+{
+	struct sbat_var *new;
+
+	new = new_entry(comp_name, comp_gen);
+	if (!new)
+		return EFI_OUT_OF_RESOURCES;
+
+	list_add_tail(&new->list, list);
+	return EFI_SUCCESS;
+}
+
+EFI_STATUS
+parse_sbat_var(list_t *entries)
+{
+	UINT8 *data = 0;
+	UINTN datasize, i;
+	EFI_STATUS efi_status;
+	INIT_LIST_HEAD(entries);
+	char delim;
+
+	if (!entries)
+		return EFI_INVALID_PARAMETER;
```

Comment:
```suggestion
	char delim;

	if (!entries)
		return EFI_INVALID_PARAMETER;

	INIT_LIST_HEAD(entries);
```
`INIT_LIST_HEAD(entries)` should be after the `!entries` sanity check.

Reply:
should be fixed.